### DISCUSSION TITLE
How to sign in with Apple and Google credentials in react-native (Without Expo)

### DISCUSSION THREAD

**Discussion by hillkim** [supabase/supabase#2072] [Questions]
I have been trying to implement sign in with google and apple using the following libraries in bare react-native projects. 

-  [react-native-apple-authentication](https://github.com/invertase/react-native-apple-authentication)

-  [react-native-google-signin](https://github.com/react-native-google-signin/google-signin)

They work very well with firebase as described [here](https://rnfirebase.io/auth/social-auth) using firebase sdk.
The concept is fairly simple, the native sdk for both apple and google sign return user credential after a successful sign in . Here is an example from [https://rnfirebase.io/auth/social-auth](https://rnfirebase.io/auth/social-auth)

### With google credential
```js
import auth from '@react-native-firebase/auth';
import { GoogleSignin } from '@react-native-google-signin/google-signin';

async function onGoogleButtonPress() {
  // Get the users ID token
  const { idToken } = await GoogleSignin.signIn();

  // Create a Google credential with the token
  const googleCredential = auth.GoogleAuthProvider.credential(idToken);

  // Sign-in the user with the credential
  return auth().signInWithCredential(googleCredential);
}
```
### With apple credential
```js
import auth from '@react-native-firebase/auth';
import { appleAuth } from '@invertase/react-native-apple-authentication';

async function onAppleButtonPress() {
  // Start the sign-in request
  const appleAuthRequestResponse = await appleAuth.performRequest({
    requestedOperation: appleAuth.Operation.LOGIN,
    requestedScopes: [appleAuth.Scope.EMAIL, appleAuth.Scope.FULL_NAME],
  });

  // Ensure Apple returned a user identityToken
  if (!appleAuthRequestResponse.identityToken) {
    throw 'Apple Sign-In failed - no identify token returned';
  }

  // Create a Firebase credential from the response
  const { identityToken, nonce } = appleAuthRequestResponse;
  const appleCredential = auth.AppleAuthProvider.credential(identityToken, nonce);

  // Sign the user in with the credential
  return auth().signInWithCredential(appleCredential);
}
```

I have gone through the auth provided by supabase here [supabase-js auth](https://supabase.io/docs/reference/javascript/auth-signin)
The closest thing I can find related to the implementation above is the use of a refresh-token as shown here 
[Sign in using a refresh token ](https://supabase.io/docs/reference/javascript/auth-signin#sign-in-using-a-refresh-token-eg-in-react-native)
```js
// An example using Expo's `AuthSession`
const redirectUri = AuthSession.makeRedirectUri({ useProxy: false });
const provider = 'google';

AuthSession.startAsync({
  authUrl: `https://MYSUPABASEAPP.supabase.co/auth/v1/authorize?provider=${provider}&redirect_to=${redirectUri}`,
  returnUrl: redirectUri,
}).then(async (response: any) => {
  if (!response) return;
  const { user, session, error } = await supabase.auth.signIn({
    refreshToken: response.params?.refresh_token,
  });
});
```
The example above uses a **refreshToken** parameter  _auth.sign()_ to create a session for a user who is already registered.

**Correct me if am wrong.** According to the auth documentation there is no method that allows one to create a user using an auth credential. This is because of how gotrue social auth is implemented. The only solution I can think of as now is to move the entire auth logic to the backend and handle it on the web. This is not elegant at all since it will have to involve ugly deep-linking or webview implementations to  use supabase session and access token.

Given the scenario above, is there way of using native social sdk  in react-native to create a new user account in supabase using a user credential without having to force them use an email link,  password or web ?   Any ideas , answers or corrections are welcome.

**Comment by jstnw10** [comment]
As far as I know you're right and there is currently no way to use a native sdk this way. Would be nice to see it soon as my flutter application is waiting on this feature too before it can migrate to supabase.

**Comment by megacherry** [comment]
I also think you are right. I doesn't work right now. I think it's the best to create an Issue and we can upvote it? What do you think about it?

**Comment by kangmingtay** [comment]
Hey @megacherry, @jstnw10, @hillkim, we are currently adding support for this! You can keep track of the open issue [here](https://github.com/supabase/gotrue/issues/140). Also, we have a WIP PR [here](https://github.com/supabase/gotrue/pull/189) which aims to improve the integration between supabase-auth and the native mobile auth frameworks.

**Reply by fikrikarim** [reply]
I'm waiting for this too. Hope it could be released to prod soon.

**Reply by megacherry** [reply]
> Hey @megacherry, @jstnw10, @hillkim, we are currently adding support for this! You can keep track of the open issue [here](https://github.com/supabase/gotrue/issues/140). Also, we have a WIP PR [here](https://github.com/supabase/gotrue/pull/189) which aims to improve the integration between supabase-auth and the native mobile auth frameworks.

It is now in production but how can we use it? Is there any documentation in the wild?

**Comment by cordial** [comment]
@kangmingtay Nothing in any of the links supplied show a way to now do this using 

 [react-native-apple-authentication](https://github.com/invertase/react-native-apple-authentication)
or
[react-native-google-signin](https://github.com/react-native-google-signin/google-signin). Are you suggesting it's now done in another way? Again the links and PR really dont offer much in the way of help for direction to do this. Are you suggesting you need to do it without the libraries mention at the top of the discussion? Regarding Apple signin, I can't find any decent documentation on this apart from a comment in supabase Login with Apple page saying if you're not using Expo, then use [react-native-apple-authentication](https://github.com/invertase/react-native-apple-authentication).

**Comment by cordial** [comment]
Ok, to anyone who reads this and needs help. I got apple signing (on iOs) working by using [react-native-apple-authentication](https://github.com/invertase/react-native-apple-authentication) by creating a raw nonce (who makes up these names) and passing it to performRequest and then taking the same rawNonce and passing it to the supabase signin afterwards.  So...

`
const rawNonce = uuidv4()
      const state = uuidv4()
      const appleAuthRequestResponse = await appleAuth.performRequest({
        requestedOperation: appleAuth.Operation.LOGIN,
        requestedScopes: [appleAuth.Scope.FULL_NAME, appleAuth.Scope.EMAIL],
        nonce: rawNonce,  
        state: state,
      })
      const credentialState = await appleAuth.getCredentialStateForUser(appleAuthRequestResponse.user)
      
       if (
        appleAuthRequestResponse.identityToken &&
        credentialState === appleAuth.State.AUTHORIZED
      ) {
        const {
          error,
          data: { user },
        } = await supabase.auth.signInWithIdToken({
          provider: 'apple',
          token: appleAuthRequestResponse.identityToken,
          nonce: rawNonce, // Pass the same raw nonce
        })
 `
